scipy-stubs/signal/_peak_finding.pyi

```diff
@@ -1,37 +1,72 @@
-from scipy._typing import Untyped
+from collections.abc import Callable, Sequence
+from typing import Concatenate, Literal, TypeAlias, TypedDict
+from typing_extensions import NotRequired
+
+import numpy as np
+import numpy.typing as npt
 
 __all__ = ["argrelextrema", "argrelmax", "argrelmin", "find_peaks", "find_peaks_cwt", "peak_prominences", "peak_widths"]
 
-def argrelmin(data: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def argrelmax(data: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def argrelextrema(data: Untyped, comparator: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def peak_prominences(x: Untyped, peaks: Untyped, wlen: Untyped | None = None) -> Untyped: ...
+_Mode: TypeAlias = Literal["clip", "wrap"]
+_ProminencesResult: TypeAlias = tuple[npt.NDArray[np.float64], npt.NDArray[np.intp], npt.NDArray[np.intp]]
+_WidthsResult: TypeAlias = tuple[
+    npt.NDArray[np.float64], npt.NDArray[np.float64], npt.NDArray[np.float64], npt.NDArray[np.float64]
+]
+
+class _FindPeaksResultsDict(TypedDict):
+    peak_heights: NotRequired[npt.NDArray[np.float64]]
+    left_thresholds: NotRequired[npt.NDArray[np.float64]]
+    right_thresholds: NotRequired[npt.NDArray[np.float64]]
+    prominences: NotRequired[npt.NDArray[np.float64]]
+    left_bases: NotRequired[npt.NDArray[np.intp]]
+    right_bases: NotRequired[npt.NDArray[np.intp]]
+    width_heights: NotRequired[npt.NDArray[np.float64]]
+    left_ips: NotRequired[npt.NDArray[np.float64]]
+    right_ips: NotRequired[npt.NDArray[np.float64]]
+    plateau_sizes: NotRequired[npt.NDArray[np.intp]]
+    left_edges: NotRequired[npt.NDArray[np.intp]]
+    right_edges: NotRequired[npt.NDArray[np.intp]]
+
+def argrelmin(
+    data: npt.NDArray[np.generic], axis: int = 0, order: int = 1, mode: _Mode = "clip"
+) -> tuple[npt.NDArray[np.intp], ...]: ...
+def argrelmax(
+    data: npt.NDArray[np.generic], axis: int = 0, order: int = 1, mode: _Mode = "clip"
+) -> tuple[npt.NDArray[np.intp], ...]: ...
+def argrelextrema(
+    data: npt.NDArray[np.generic],
+    comparator: Callable[[npt.NDArray[np.generic], npt.NDArray[np.generic]], npt.NDArray[np.bool_]],
+    axis: int = 0,
+    order: int = 1,
+    mode: _Mode = "clip",
+) -> tuple[npt.NDArray[np.intp], ...]: ...
+def peak_prominences(x: npt.ArrayLike, peaks: npt.ArrayLike, wlen: int | None = None) -> _ProminencesResult: ...
 def peak_widths(
-    x: Untyped,
-    peaks: Untyped,
+    x: npt.ArrayLike,
+    peaks: npt.ArrayLike,
     rel_height: float = 0.5,
-    prominence_data: Untyped | None = None,
-    wlen: Untyped | None = None,
-) -> Untyped: ...
+    prominence_data: _ProminencesResult | None = None,
+    wlen: int | None = None,
```

Comment:
The docs are incorrect here; `float` is also valid as `wlen` (in which case `math.ceil` is used).
And apparently, passing `wlen=np.uint8(4)` and even `wlen=np.float16(4)` is also allowed, so `scipy._typing.AnyReal` could be used for this.

Allowing this to be a float is a bit weird IMHO...
But if it's allowed; it's allowed 🤷🏻

Reply:
I actually remember seeing the `math.ceil` but I forgot about it when writing the type annotation.

```python
# Round up to a positive integer
if isinstance(value, float):
    value = math.ceil(value)
value = np.intp(value)
```

I feel like this is a bit of a bug or just inefficient code. I'm pretty sure np.floatx doesn't pass the isinstance check which means it gets converted to np.intp which doesn't perform a ceil like it does for floats but if that is fine then there is no need to single out `float`.

As a type annotation though I will use `AnyReal` because it is allowed technically.